src/scan/block_processor.rs

```diff
@@ -0,0 +1,550 @@
+    }
+}
+
+struct BlockEventAccumulator {
+    account_id: i64,
+    height: u64,
+    block_hash: Vec<u8>,
+    outputs: Vec<DetectedOutput>,
+    inputs: Vec<SpentInput>,
+    confirmations: Vec<ConfirmedOutput>,
+    full_balance_changes: Vec<BalanceChange>,
+}
+
+impl BlockEventAccumulator {
+    fn new(account_id: i64, height: u64, block_hash: Vec<u8>) -> Self {
+        Self {
+            account_id,
+            height,
+            block_hash,
+            outputs: Vec::new(),
+            inputs: Vec::new(),
+            confirmations: Vec::new(),
+            full_balance_changes: Vec::new(),
+        }
+    }
+
+    fn add_balance_change(&mut self, change: BalanceChange) {
+        self.full_balance_changes.push(change);
+    }
+
+    fn build(self) -> BlockProcessedEvent {
+        let balance_changes = self
+            .full_balance_changes
+            .iter()
+            .map(|c| BalanceChangeSummary {
+                credit: c.balance_credit,
+                debit: c.balance_debit,
+                description: c.description.clone(),
+            })
+            .collect();
+
+        BlockProcessedEvent {
+            account_id: self.account_id,
+            height: self.height,
+            block_hash: self.block_hash,
+            outputs_detected: self.outputs,
+            inputs_spent: self.inputs,
+            outputs_confirmed: self.confirmations,
+            balance_changes,
+        }
+    }
+}
+
+struct MemoInfo {
+    parsed: Option<String>,
+    hex: Option<String>,
+}
+
+impl MemoInfo {
+    fn from_output(output: &WalletOutput) -> Self {
+        let payment_info = output.payment_id();
+        let memo_bytes = payment_info.get_payment_id();
+
+        if memo_bytes.is_empty() {
+            Self {
+                parsed: None,
+                hex: None,
+            }
+        } else {
+            Self {
+                parsed: Some(String::from_utf8_lossy(&memo_bytes).to_string()),
+                hex: Some(hex::encode(&memo_bytes)),
+            }
+        }
+    }
+}
+
+fn make_balance_change_for_output(
+    account_id: i64,
+    output_id: i64,
+    timestamp: u64,
+    height: u64,
+    output: &WalletOutput,
+) -> BalanceChange {
+    let effective_date = DateTime::<Utc>::from_timestamp(timestamp as i64, 0)
+        .unwrap_or_else(Utc::now)
+        .naive_utc();
+
+    if output.features().is_coinbase() {
+        return BalanceChange {
+            account_id,
+            caused_by_output_id: Some(output_id),
+            caused_by_input_id: None,
+            description: "Coinbase output found in blockchain scan".to_string(),
+            balance_credit: output.value().as_u64(),
+            balance_debit: 0,
+            effective_date,
+            effective_height: height,
+            claimed_recipient_address: None,
+            claimed_sender_address: None,
+            memo_parsed: None,
+            memo_hex: None,
+            claimed_fee: None,
+            claimed_amount: None,
```

Comment:
these you should be able to fill in, a coinbase has a memop field, and an amount, the fee is always 0 for coinbase